File: Semestrul 4/MPP/lab09-10/server .net/TriathlonServer/server/ServiceImpl.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Collections.Concurrent;
using log4net;
using Triathlon.Persistence.Interfaces;
using Triathlon.Model;
using Triathlon.Services;

namespace Triathlon.Server
{
    public class ServiceImplementation : ITriathlonServices
    {
        private readonly IEventRepository _eventRepository;
        private readonly IParticipantRepository _participantRepository;
        private readonly IRefereeRepository _refereeRepository;
        private readonly IResultRepository _resultRepository;
        private readonly IRefereeDisciplineRepository _refereeDisciplineRepository;

                 private ConcurrentDictionary<long, ITriathlonObserver> _loggedReferees;

                 private static readonly ILog _logger = LogManager.GetLogger(typeof(ServiceImplementation));

                 private readonly int _defaultThreadsCount = 5;

        public ServiceImplementation(
            IEventRepository eventRepository,
            IParticipantRepository participantRepository,
            IRefereeRepository refereeRepository,
            IResultRepository resultRepository,
            IRefereeDisciplineRepository refereeDisciplineRepository)
        {
            _eventRepository = eventRepository;
            _participantRepository = participantRepository;
            _refereeRepository = refereeRepository;
            _resultRepository = resultRepository;
            _refereeDisciplineRepository = refereeDisciplineRepository;
            _loggedReferees = new ConcurrentDictionary<long, ITriathlonObserver>();
        }

        public RefereeDiscipline Login(Referee tempReferee, ITriathlonObserver client)
        {
            lock (this)
            {
                _logger.Info($"Login attempt from referee: {tempReferee.Username}");

                var refereeOpt = _refereeRepository.FindByUsername(tempReferee.Username);
                if (refereeOpt == null)
                {
                    _logger.Error($"Referee not found: {tempReferee.Username}");
                    throw new TriathlonException("Invalid username or password.");
                }

                var refereeDisciplineOpt = _refereeDisciplineRepository.FindByRefereeId(refereeOpt.Id);
                if (refereeDisciplineOpt != null)
                {
                    Referee existingReferee = refereeOpt;

                    if (existingReferee.Password.Equals(tempReferee.Password))
                    {
                        if (_loggedReferees.ContainsKey(existingReferee.Id))
                        {
                            _logger.Error($"Referee already logged in: {existingReferee.Username}");
                            throw new TriathlonException("Referee already logged in.");
                        }

                        _loggedReferees[existingReferee.Id] = client;
                        _logger.Info($"Referee logged in successfully: {existingReferee.Username}");

                        return refereeDisciplineOpt;
                    }
                    else
                    {
                        _logger.Error($"Invalid password for referee: {existingReferee.Username}");
                        throw new TriathlonException("Invalid username or password.");
                    }
                }
                else
                {
                    _logger.Error($"Referee doesn't have assigned any discipline: {refereeOpt.Username}");
                    throw new TriathlonException("Invalid username or password.");
                }
            }
        }

        public void Logout(Referee referee, ITriathlonObserver client)
        {
            lock (this)
            {
                _logger.Info($"Logout attempt from referee: {referee.Username}");

                if (!_loggedReferees.TryRemove(referee.Id, out ITriathlonObserver localClient))
                {
                    _logger.Error($"Referee not logged in: {referee.Username}");
                    throw new TriathlonException("Referee is not logged in.");
                }

                _logger.Info($"Referee logged out successfully: {referee.Username}");
            }
        }

        public List<Result> GetResultsByEvent(long eventId)
        {
            _logger.Info($"Getting results for event ID: {eventId}");
            return _resultRepository.FindByEventId(eventId).ToList();
        }

        public void UpdateResult(Result result)
        {
            lock (this)
            {
                _logger.Info($"Updating result ID: {result.Id} for participant: {result.Participant.Id}");

                _resultRepository.Update(result);
                NotifyResultUpdated(result);
            }
        }

        public int GetTotalPointsForParticipant(long participantId, long eventId)
        {
            return _resultRepository.FindByParticipantId(participantId)
                .Where(r => r.Event.Id.Equals(eventId))
                .Sum(r => r.Points);
        }

        private void NotifyResultUpdated(Result result)
        {
            _logger.Info($"Notifying result update for result ID: {result.Id}");

                         List<Thread> threads = new List<Thread>();
            _logger.Info($"There are {_loggedReferees.Count} clients connected");
            foreach (var entry in _loggedReferees)
            {
                var observerId = entry.Key;
                var observer = entry.Value;
                _logger.Info($"Notifying observer with ID: {observerId}");

                Thread thread = new Thread(() =>
                {
                    try
                    {
                        _logger.Debug($"Notifying referee ID: {observerId} about result update");
                        observer.ResultsUpdated(result);
                    }
                    catch (TriathlonException ex)
                    {
                        _logger.Error($"Error notifying referee: {ex.Message}");
                    }
                });

                threads.Add(thread);
                thread.Start();
            }

            
        }
    }
}

File: Semestrul 4/MPP/lab07/java/TriathlonServer/src/main/java/server/ServiceImpl.java
package server;

import triathlon.persistence.interfaces.*;
import triathlon.model.*;
import triathlon.services.ITriathlonObserver;
import triathlon.services.ITriathlonServices;
import triathlon.services.TriathlonException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ServiceImpl implements ITriathlonServices {
    private final EventRepository eventRepository;
    private final ParticipantRepository participantRepository;
    private final RefereeRepository refereeRepository;
    private final ResultRepository resultRepository;
    private final RefereeDisciplineRepository refereeDisciplineRepository;


    private Map<Long, ITriathlonObserver> loggedReferees;


    private static final Logger logger = LogManager.getLogger(ServiceImpl.class);


    private final int defaultThreadsNo = 5;

    public ServiceImpl(EventRepository eventRepository, ParticipantRepository participantRepository,
                       RefereeRepository refereeRepository, ResultRepository resultRepository,
                       RefereeDisciplineRepository refereeDisciplineRepository) {
        this.eventRepository = eventRepository;
        this.participantRepository = participantRepository;
        this.refereeRepository = refereeRepository;
        this.resultRepository = resultRepository;
        this.refereeDisciplineRepository = refereeDisciplineRepository;
        this.loggedReferees = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized RefereeDiscipline login(Referee tempReferee, ITriathlonObserver client) throws TriathlonException {
        logger.info("Login attempt from referee: " + tempReferee.getUsername());

        Optional<Referee> refereeOpt = refereeRepository.findByUsername(tempReferee.getUsername());
        if (refereeOpt.isEmpty()) {
            logger.error("Referee not found: " + tempReferee.getUsername());
            throw new TriathlonException("Invalid username or password.");
        }

        var refereeDisciplineOpt = refereeDisciplineRepository.findByRefereeId(refereeOpt.get().getId());
        if (refereeDisciplineOpt.isPresent()) {
            Referee existingReferee = refereeOpt.get();

            if (existingReferee.getPassword().equals(tempReferee.getPassword())) {
                if (loggedReferees.containsKey(existingReferee.getId())) {
                    logger.error("Referee already logged in: " + existingReferee.getUsername());
                    throw new TriathlonException("Referee already logged in.");
                }

                loggedReferees.put(existingReferee.getId(), client);
                logger.info("Referee logged in successfully: " + existingReferee.getUsername());

                return refereeDisciplineOpt.get();
            } else {
                logger.error("Invalid password for referee: " + existingReferee.getUsername());
                throw new TriathlonException("Invalid username or password.");
            }
        } else {
            logger.error("Referee doesn't have assigned any discipline: " + refereeOpt.get().getUsername());
            throw new TriathlonException("Invalid username or password.");
        }
    }

    @Override
    public synchronized void logout(Referee referee, ITriathlonObserver client) throws TriathlonException {
        logger.info("Logout attempt from referee: " + referee.getUsername());

        ITriathlonObserver localClient = loggedReferees.remove(referee.getId());

        if (localClient == null) {
            logger.error("Referee not logged in: " + referee.getUsername());
            throw new TriathlonException("Referee is not logged in.");
        }

        logger.info("Referee logged out successfully: " + referee.getUsername());
    }

    @Override
    public List<Result> getResultsByEvent(Long eventId) {
        logger.info("Getting results for event ID: " + eventId);
        return resultRepository.findByEvent(eventId.intValue());
    }

    @Override
    public synchronized void updateResult(Result result) {
        logger.info("Updating result ID: " + result.getId() + " for participant: " + result.getParticipant().getId());

        resultRepository.update(result);
        notifyResultUpdated(result);
    }

    @Override
    public int getTotalPointsForParticipant(Long participantId, Long eventId) {
        return resultRepository.findByParticipant(Math.toIntExact(participantId)).stream()
                .filter(r -> r.getEvent().getId().equals(eventId))
                .mapToInt(Result::getPoints)
                .sum();
    }

    private void notifyResultUpdated(Result result) {
        logger.info("Notifying result update for result ID: " + result.getId());
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);
        for (var entry : loggedReferees.entrySet()) {
            var observer = entry.getValue();
            executor.execute(() -> {
                try {
                    logger.debug("Notifying referee ID: " + entry.getKey() + " about result update");
                    observer.resultsUpdated(result);
                } catch (TriathlonException e) {
                    logger.error("Error notifying referee: " + e.getMessage());
                }
            });
        }
        executor.shutdown();
    }
}





